tfutils/optimizer.py

```diff
@@ -144,54 +215,119 @@ def __init__(self, optimizer, *optimizer_args, **optimizer_kwargs):
+        else:
+            curr_var_list = self.var_list[opt_idx]
+
+        assert len(curr_mb_gv) == len(curr_var_list), \
                 "Variable list length not matched!"
         assert all((\
                 var_g.name == var_l.name \
-                for (_, var_g), var_l in zip(minibatch_grads, self.var_list))),\
+                for (_, var_g), var_l in zip(curr_mb_gv, curr_var_list))),\
                 "Variable list should have the same variables!"
-        if self.grads_and_vars is None:
-            self.grads_and_vars = [(
-                tf.Variable(
-                    tf.zeros_like(var.initialized_value()),
-                    dtype=tf.float32,
-                    trainable=False,
-                    name=NON_SAVE_SUFFIX,
-                    ),
-                var) for var in self.var_list]
 
+    def _zero_gvs(self, opt_idx=None):
+        if opt_idx is None:
+            curr_var_list = self.var_list
+        else:
+            curr_var_list = self.var_list[opt_idx]
+
+        zero_gvs = [(
+                    tf.Variable(
+                        tf.zeros_like(var.initialized_value()),
+                        dtype=tf.float32,
+                        trainable=False,
+                        name=NON_SAVE_SUFFIX,
+                        ),
+                    var) for var in curr_var_list]
+        return zero_gvs
+
+    def _mini_ops(self, grads_and_vars, minibatch_grads, num_minibatches=1):
         mini_ops = []
         for (grad_v, _), (mini_grad, _) \
-                in zip(self.grads_and_vars, minibatch_grads):
+                in zip(grads_and_vars, minibatch_grads):
             mini_ops.append(
                     tf.assign_add(grad_v, mini_grad / num_minibatches))
-        mnb_accu_grad = tf.group(*mini_ops)
+        return mini_ops
+
+    def accumulate_gradients(self, minibatch_grads, num_minibatches=1):
+        """Accumulate gradients for `num_minibatches` minibatches."""
+        if num_minibatches == 1:
+            # No need for accumulating
+            return tf.no_op(), minibatch_grads
+
+        if self._multi_mode: # list of lists, grads and vars per optimizer
+            assert(len(minibatch_grads) == len(self._optimizer._optimizer_class))
+            for opt_idx, curr_mb_gv in enumerate(minibatch_grads):
+                self._consistency_check(curr_mb_gv, opt_idx=opt_idx)
+
+            if self.grads_and_vars is None:
+                zero_gvs_list = []
+                for opt_idx, _ in enumerate(minibatch_grads):
+                    zero_gvs_list.insert(opt_idx, self._zero_gvs(opt_idx=opt_idx))
+                self.grads_and_vars = zero_gvs_list
+
+            mini_ops_list = []
+            for opt_idx, curr_mb_gv in enumerate(minibatch_grads):
+                curr_mini_ops = self._mini_ops(grads_and_vars=self.grads_and_vars[opt_idx], minibatch_grads=curr_mb_gv, num_minibatches=num_minibatches)
+                mini_ops_list.extend(curr_mini_ops)
```

Comment:
I noticed here `extend` is used to append to the list, whereas in https://github.com/neuroailab/tfutils/pull/146/files#diff-51c7e10eb689dc78b0e9fa24df942647R306 (and elsewhere, when looping over optimizers) `insert` is used. I think this should be consistent throughout the optimizer.